<summary><b>Input type</b></summary>
<summary><ins>Folder</ins></summary>

User must organize an input folder containing all of the vcf and tsv files requested following the structure reported below:

```
input_folder/
├── CNV
│   ├── 001.vcf
│   ├── 002.vcf
│   └── 003.vcf
├── SNV
│   ├── 001.vcf
│   ├── 002.vcf
│   └── 003.vcf
├── CombinedOutput
│   ├── 001_CombinedVariantOutput.tsv
│   ├── 002_CombinedVariantOutput.tsv
│   └── 003_CombinedVariantOutput.tsv
├── FUSIONS
│   └── Fusions.tsv
├── sample.tsv   
└── patient.tsv
```

Where:

* <b>SNV</b> folder contains all vcf files related to single nucleotide variants.
* <b>CNV</b> folder contains all vcf files related to copy number variants.
* <b>CombinedVariantOutput</b> folder contains all the Combined Variant Output file in tsv format (this kind of files contains info about TMB, MSI and Fusions).
* <b>FUSIONS</b> folder contains the template fusion.tsv, which can filled with fusion data if the Combined Variant Output files are unavailable.
* <b>sample.tsv</b> file is a template for clinical data of the samples.
* <b>patient.tsv</b> file is a template for clinical data of the patients.



<summary><ins>File(s)</ins></summary>
User must compile several input file (by filling in specific templates):

* <b>sample.tsv</b> file, a template to fill with samples' clinical info.
* <b>fusions.tsv</b> file, a template to fill with fusions' information.
* <b>patient.tsv</b> file, a template to fill with patients' clinical info.

⚠️ For this input, only the <i>sample.tsv</i> file is mandatory.

<br>

  <summary><b>Templates</b></summary>

For both input by folder and by file, template filling by user is requested.
Below will be briefly explained the structure of these templates:


  <summary><i>sample.tsv</i></summary>
This template must be filled by user with all disposable samples' clinical info and will be used to create the <i>data_clinical_sample.txt</i>.

⚠️ This file is mandatory for Varan analysis!

|SAMPLE_ID | PATIENT_ID | MSI | TMB| MSI_THR | TMB_THR| ONCOTREE_CODE| snv_path| cnv_path| comb_path|... |
|:---:|:---:|:---: |:---:   |:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|0000000_DNA| 00000000  | 1| 12| ||BOWEL |path_to_snv|path_to_cnv|path_to_combined_output|...|
|0000001_DNA| 00000001  | 8.0| 8.0| ||UTERUS |path_to_snv|path_to_cnv|path_to_combined_output|...|
|0000003_DNA| 00000003  | 222| 127| ||BOWEL |path_to_snv|path_to_cnv|path_to_combined_output|...|
|...| ...  | ...| ...| ||... |...|...|...|...|

The obligatory fields to keep are:

* <ins>SAMPLE_ID</ins>: IDs of all samples of interest.
* <ins>PATIENT_ID</ins>: IDs of all patient of interest.
* <ins>MSI</ins>: MSI value for each of the samples.
* <ins>TMB</ins>: TMB value for each of the samples.
* <ins>MSI_THR</ins>: MSI categorization based on the threshold set in the <i>conf.ini</i> file. This field has to be left empty and will be filled by Varan.
* <ins>TMB_THR</ins>: TMB categorization based on the threshold set in the <i>conf.ini</i> file. This field has to be left empty and will be filled by Varan.
* <ins>ONCOTREE_CODE</ins>: code to associate for the oncokb annotation. Check <a href="https://oncotree.mskcc.org/#/home">here</a> for more info.
* <ins>snv_path</ins>: path to the SNV vcf file. This column has to be filled in case of input by file.
* <ins>cnv_path</ins>: path to the CNV vcf file. This column has to be filled in case of input by file.
* <ins>comb_path</ins>: path to the Combined Variant Output file. This column has to be filled in case of input by file.

⚠️ The user can add new columns starting from the last one. Modifying or deleting the default ones (even only by changing names) can lead to errors and is strongly not recommended.<br>
⚠️ Both SAMPLE_ID and PATIENT_ID must be filled.<br>
⚠️ The MSI and TMB columns will be read from this file only if no Combined Variant Output file is provided. If the Combined Variant Output files are available, those values will be prioritized.<br>
⚠️ ONCOTREECODE column is mandatory to fill for the oncoKB annotation


  <summary><i>patient.tsv</i></summary>

This template must be filled by user with all disposable patients' clinical info and will be used to create the <i>data_clinical_patient.txt</i>.

⚠️ This file is optional. If missing, a default <i>data_clinical_patient.txt</i> will be create.


|PATIENT_ID | AGE | SEX | ...|
|:---:|:---:|:---:|:---:|
|00000000| 45  | F| ...| 
|00000001| 77  | F| ...| 
|00000003| 23  | M| ...| 
|...| ...  | ...|...| 

The obligatory fields to keep are:
* <ins>PATIENT_ID</ins>: IDs of all patient of interest

⚠️ The user can add new columns starting from PATIENT_ID. Modifying or deleting the default one (even only by changing name) can lead to errors and is strongly not recommended.<br>


  <summary><i>fusions.tsv</i></summary>

This template must be filled by user with all disposable fusion info and will be used to create the <i>data_sv.txt</i>.

⚠️ This file is optional.

|Sample_Id | SV_Status | Site1_Hugo_Symbol |Site2_Hugo_Symbol|...|
|:---:|:---:|:---:|:---:|:---:|
|0000000_DNA| SOMATIC  | APC| BRCA1|...|
|0000001_DNA| SOMATIC  | TP53| BRAF|...|
|0000003_DNA| SOMATIC  | ALK| BRCA2|...|
|...| ... | ...|...|...|

The obligatory fields to keep are:
* <ins>Sample_Id</ins>: IDs of all sample of interest
* <ins>SV_Status</ins>: fusion type
* <ins>Site1_Hugo_Symbol</ins>: first gene involved in fusion 
* <ins>Site2_Hugo_Symbol</ins>: second gene involved in fusion 

⚠️ The user can add new columns starting from Site2_Hugo_Symbol. Modifying or deleting the default ones (even only by changing names) can lead to errors and is strongly not recommended.<br>
⚠️ All mandatory fields must be filled and cannot be left empty.<br>



<details open>
  <summary><b>Configuration file</b></summary>
The first step to start using Varan is to correctly set the configuration file <i>conf.ini</i>. <br><br>

⚠️ In the next subparagraph, for each field the type of variable requested will be insert between angle brackets <>. For string two possible entry can be find: < 'string' > and < string >. In the first case it is request to insert text inside quotation marks (*i.e. DESCRIPTION='this is the description'*), while on the other one quotation marks are not requested (*i.e. PROJECT_NAME=study*).

This file is divided in 12 subsessions:
<details close>
<summary><ins>Paths</ins></summary>
In this section it's possible to specify the paths for Vep (VEP_PATH) and its cache (VEP_DATA) and fasta (REF_FASTA), vcf2maf (VCF2MAF), and ClinVar (CLINV). In CACHE it is possible to set Ensembl version (i.e. 111)

```
[Annotations]
ANNOTATIONS = < '' >
```
⚠️ clinvar database can be downloaded <a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar/">here</a>

</details>
<details close>
<summary><ins>Multiple</ins></summary>

In this section it's possible to specify the paths in case of multiple SNV, CNV and/or Combined Variant Output files analysis.   
```
[Multiple]
SNV= < string >
CNV= < string >
COMBOUT= < string >
```
⚠️ This section has to be filled only in case of input by file.

</details>
<details close>
<summary><ins>Zip</ins></summary>

In this section you can decide how to manage MAF and SNV_FILTERED files.
```
[Zip]
ZIP_MAF = < boolean >
ZIP_SNV_FILTERED = < boolean >
COPY_MAF = < boolean >
```
</details>
<details close>
<summary><ins>OncoKB</ins></summary>

In this section is possible to insert the personal <a href="https://www.oncokb.org/">oncoKB</a> key. This key is mandatory to execute the oncoKB annotation.

```
[OncoKB] ONCOKB
ONCOKB= < string >
```
⚠️ The request for the oncoKB key can be done <a href="https://www.oncokb.org/account/register">here</a>


</details>
<details close>
<summary><ins>Project</ins></summary>

In this section is possible to specify project info like study name, ID, description and profile. These info will be insert in meta files.
```
[Project]
PROJECT_ID = < string >
PROJECT_NAME = < string >
DESCRIPTION = < 'string' >
PROFILE_MUT = < 'string' >
PROFILE_CNA = < 'string' >
PROFILE_CNA_HG19 = < 'string' >
PROFILE_SV = < 'string' >
```

</details>
<details close>
<summary><ins>Filters</ins></summary>

Here it is possible to specify the filters' threshold to apply to SNV maf files (for more info about filters threshold setting see [Ex 7)](#ex-7-filter-vcfmaf).
```
[Filters]
CLIN_SIG = < ['string','string',...] >
CONSEQUENCES = < ['string','string',...] >
ONCOKB_FILTER = < ['string','string',...] >
t_VAF_min = < float >
t_VAF_min_novel = < float >
t_VAF_max = < float >
AF = < string > 
POLYPHEN = < ['string','string',...] >
IMPACT = < ['string','string',...] >
SIFT = < ['string','string',...] >
drop_NA_AF = < boolean >
```
⚠️ For AF the field can be populate with </>/<=/>=val (i.e. AF = <0.003)

</details close>
<details>
<summary><ins>Cna</ins></summary>

In this section user can insert CNV genotypes of interest and ploidy. The latter will be used to evaluate copy number discretization using <a href="https://cnvkit.readthedocs.io/en/stable/pipeline.html">cnvkit formula</a>.
```
[Cna]
HEADER_CNV = < ['string','string',...] >
PLOIDY = < int >
CNVkit = < boolean >
```
</details close>
<details>
<summary><ins>TMB</ins></summary>

Here TMB thresholds can be specified.

```
[TMB]
THRESHOLD_TMB = < {'string':'string', 'string':'string', ...} > 
```
i.e. THRESHOLD = {'Low':'<=5','Medium':'<10','High':'>=10'} where the string before : is the label assign to TMB value, while the other is the specific threshold (i.e. for a sample with TMB=15 a label 'High' will be reported in the data clinical sample).

</details close>
<details>
<summary><ins>MSI</ins></summary>

Here MSI thresholds for sites and values can be specified.
```
[MSI]
THRESHOLD_SITES = < string >
THRESHOLD_MSI = < string >
```

⚠️ THRESHOLD_SITES value will be used only if MSI informations are extracted from Combined Variant Output files. If MSI is directly reported as a value inside the input tsv, only THRESHOLD_MSI will be applied.

⚠️ Both THRESHOLD_SITES and THRESHOLD_MSI can be populate with </>/<=/>=val (i.e. THRESHOLD = <20).

</details close>
<details>
<summary><ins>FUSION</ins></summary>

Here Fusions thresholds can be specified.

```
[FUSION]
THRESHOLD_FUSION = < string >
```
⚠️ THRESHOLD can be populate with </>/<=/>=val (i.e. THRESHOLD = >=15)
</details>
<details close>
<summary><ins>ClinicalSample</ins></summary>

Users can customize column names and data types for the data_clinical_sample.txt file.
```
[ClinicalSample]
HEADER_SAMPLE_SHORT = < ['string','string',...] >
HEADER_SAMPLE_LONG = < ['string','string',...] >
HEADER_SAMPLE_TYPE = < ['string','string',...] >
```
⚠️ HEADER_SAMPLE_TYPE accept only STRING, NUMBER, BOOLEAN. If different type is inserted, an error will be raised by Varan.

⚠️ If these fields are left empty, a default Header will be produced.

</details>
<details close>
<summary><ins>ClinicalPatient</ins></summary>

Users can customize column names and data types for the data_clinical_patient.txt file.
```
[ClinicalPatient]
HEADER_PATIENT_SHORT = < ['string','string',...] >
HEADER_PATIENT_LONG = < ['string','string',...] >
HEADER_PATIENT_TYPE = < ['string','string',...] >
```
⚠️ HEADER_PATIENT_TYPE accept only STRING, NUMBER, BOOLEAN. If different type is inserted, an error will be raised by Varan.

⚠️ If these fields are left empty, a default Header will be produced.

</details>
<details close>
<summary><ins>Annotations</ins></summary>
In this section it's possible to insert manual notes that will appear in report_VARAN.html file.

```
[Annotations]
ANNOTATIONS = < ['string','string',...] >
```

</details>
<details close>
<summary><ins>Validation</ins></summary>

If the user has a working cbioportal instance active on his computer, the location (http://localhost:8080) can be insert here. This value will be used for the validation of the output study and will produce a html report with the results.
```
[Validation]
PORT = < string >
```
⚠️ If this field is left empty an offline validation will be conducted.
</details>

⚠️ For Docker a partially compiled configuration file (<i>docker.ini</i>) is available with vep and clinvar path already set.
</details>
